---
title: "WT and Sox2 cKO Regeneration Differential Expression"
author: "Russell Fletcher and Diya Das"
date: '`r Sys.Date()`'
output:
html_document:
code_folding: hide
toc: yes
toc_float: yes
fig_width: 10
fig_height: 10
---

```{r options, echo=FALSE, results="hide",mesasge=FALSE, error=FALSE,include=FALSE}
knitr::opts_chunk$set(fig.align="center", cache=TRUE, error=FALSE, message=FALSE, warning=FALSE, results="hide")
library(clusterExperiment); library(RColorBrewer); library(limma); library(edgeR); library(xlsx)
NMF::nmf.options(grid.patch=TRUE)

expt_str <- "Expt3"
esh <- gsub("Expt","E",expt_str)
clust_dir <- file.path("../output/clust", expt_str)
viz_dir <- file.path("../output/viz", expt_str)
DE_dir <- file.path("../output/DE", expt_str)

load(file=file.path(clust_dir, paste0(esh, "_none_fq_ruv1_nobio_nobatch_hier_merge1Kvarlf018.Rda")))

###cluster 10 equals the FABP5 cluster in the merged data
cmobjM <- cmobjM[,!(primaryClusterNamed(cmobjM) %in% c("m10"))]
###cluster 14 equals the FABP5 cluster in the ummerged data
cmobj <- cm3[,!(primaryClusterNamed(cm3) %in% c("c14"))]

#####----NOTES: cm3 (unmerged); cmobjM (merged); cmobj2M (merged, no -1s)

col.pal <- c(brewer.pal(8,"Dark2")[-c(2,3,5)],brewer.pal(12,"Paired")[c(1,2,8,10,9)],brewer.pal(12,"Set3")[c(7,8,12)], brewer.pal(8, "Pastel2")[8], brewer.pal(11,"BrBG")[11], brewer.pal(11,"PiYG")[1], "cyan", "darkblue","darkorchid2", "brown1", "springgreen1", "deepskyblue4", "darkolivegreen","antiquewhite2","azure", "cornflowerblue","blue","chartreuse","darkolivegreen3","darkorange3","deeppink4","darkred","darksalmon")
wtcols <- c(brewer.pal(8, "Blues")[-c(1:2)])
kocols <- c(brewer.pal(6, "Reds"))
cole <- c(wtcols, kocols)

```

###DE on merged clusters
```{r differentialExpression}

###########################-----ON MERGED CLUSTERS (not considering genotype), cmobjM (FABP5+ cluster (cl10) removed)

#####-----first with limma on log2 transformed post-SCONE normalized counts
##--one versus all comparisons
counts <- assay(cmobjM)
counts_log <- transform(cmobjM)
clus.labels <- factor(primaryCluster(cmobj))
clus.labelsM <- factor(primaryCluster(cmobjM))
names(clus.labels) <- names(clus.labelsM) <- colnames(counts) 

getBestNFeatures <- function(counts, cluster, contrastType, n, isCount=FALSE, exstr=NULL){
  feat <- getBestFeatures(counts, cluster, contrastType=contrastType, number=n, contrastAdj="All", p.value=0.05, isCount=isCount)
  adj <- ifelse(isCount, "voom", "limma")
  
  filename <- ifelse(is.null(exstr), paste0(esh, "_", contrast, n, "_", adj, ".txt"), paste0(esh, "_", exstr, "_", contrast, n, "_", adj, ".txt"))
  
  write.table(feat, file.path(DE_dir, filename), quote=FALSE,sep="\t")
  return(feat)
}

oneVallDE500 <- getBestNFeatures(counts_log, clus.labelsM, "OneAgainstAll", 500)
oneVallDE1000 <- getBestNFeatures(counts_log, clus.labelsM, "OneAgainstAll", 1000)
oneVallDEall <- getBestNFeatures(counts_log, clus.labelsM, "OneAgainstAll", Inf)

pairsDE500 <- getBestNFeatures(counts_log, clus.labelsM, "Pairs", 500)
pairsDE1000 <- getBestNFeatures(counts_log, clus.labelsM, "Pairs", 1000)
pairsDEall <- getBestNFeatures(counts_log, clus.labelsM, "Pairs", Inf)

##-- selecting unique DE genes
E3_oneVall500 <- unique(oneVallDE500$Feature)
E3_oneVallPairs500 <- unique(c(oneVallDE500$Feature,pairsDE500$Feature))

#####-----now with limma voom on post-SCONE normalized counts; note that I add 0.5 and voom will add 0.5 more - otherwise, there are issues with normalized counts with values < -0.5.

oneVallDE500v <- getBestNFeatures(counts+0.5, clus.labelsM, "OneAgainstAll", 500, isCount = TRUE)
oneVallDE1000v <- getBestNFeatures(counts+0.5, clus.labelsM, "OneAgainstAll", 1000, isCount = TRUE)
oneVallDEallv <- getBestNFeatures(counts+0.5, clus.labelsM, "OneAgainstAll", Inf, isCount = TRUE)

pairsDE500v <- getBestNFeatures(counts+0.5, clus.labelsM, "Pairs", 500, isCount = TRUE)
pairsDE1000v <- getBestNFeatures(counts+0.5, clus.labelsM, "Pairs", 1000, isCount = TRUE)
pairsDEallv <- getBestNFeatures(counts+0.5, clus.labelsM, "Pairs", Inf, isCount = TRUE)

##-- selecting unique DE genes
E3_oneVall500V <- unique(oneVallDE500v$Feature)
E3_oneVallPairs500V <- unique(c(oneVallDE500v$Feature,pairsDE500v$Feature))


save(oneVallDEall, oneVallDE1000, oneVallDE500, file = file.path(DE_dir, paste0(esh, "_oneVall_limmaDE.Rda")))
save(oneVallDEallv, oneVallDE1000v, oneVallDE500v, file = file.path(DE_dir, paste0(esh, "_oneVall_voomDE.Rda")))
save(pairsDEall, pairsDE1000, pairsDE500, file = file.path(DE_dir, paste0(esh, "_pairWise_limmaDE.Rda")))
save(pairsDEallv, pairsDE1000v, pairsDE500v, file = file.path(DE_dir, paste0(esh, "_pairWise_voomDE.Rda")))


save(E3_oneVall500, E3_oneVallPairs500, file= file.path(DE_dir, paste0(esh, "_limmaDEgeneLists.Rda")))
save(E3_oneVall500V, E3_oneVallPairs500V, file= file.path(DE_dir, paste0(esh, "_voomDEgeneLists.Rda")))

```

###DE on unmerged clusters
```{r differentialExpression2}

###########################-----ON UNMERGED CLUSTERS (not considering genotype), cm3B (FABP5+ cluster (cl14) removed)

#####-----first with limma on log2 transformed post-SCONE normalized counts
##--one versus all comparisons

oneVallDE500u <- getBestNFeatures(counts_log, clus.labels, "OneAgainstAll", 500, exstr = "unMerged")
oneVallDE1000u <- getBestNFeatures(counts_log, clus.labels, "OneAgainstAll", 1000, exstr = "unMerged")
oneVallDEallu <- getBestNFeatures(counts_log, clus.labels, "OneAgainstAll", Inf, exstr = "unMerged")

pairsDE500u <- getBestNFeatures(counts_log, clus.labels, "Pairs", 500, exstr = "unMerged")
pairsDE1000u <- getBestNFeatures(counts_log, clus.labels, "Pairs", 1000, exstr = "unMerged")
pairsDEallu <- getBestNFeatures(counts_log, clus.labels, "Pairs", Inf, exstr = "unMerged")

##-- selecting unique DE genes
E3_unMerged_oneVall500 <- unique(oneVallDE500u$Feature)
E3_unMerged_oneVallPairs500 <- unique(c(oneVallDE500u$Feature,pairsDE500u$Feature))

#####-----now with limma voom on post-SCONE normalized counts; note that I add 0.5 and voom will add 0.5 more - otherwise, there are issues with normalized counts with values < -0.5.

oneVallDE500vu <- getBestNFeatures(counts+0.5, clus.labels, "OneAgainstAll", 500, isCount = TRUE, exstr = "unMerged")
oneVallDE1000vu <- getBestNFeatures(counts+0.5, clus.labels, "OneAgainstAll", 1000, isCount = TRUE, exstr = "unMerged")
oneVallDEallvu <- getBestNFeatures(counts+0.5, clus.labels, "OneAgainstAll", Inf, isCount = TRUE, exstr = "unMerged")

pairsDE500vu <- getBestNFeatures(counts+0.5, clus.labels, "Pairs", 500, isCount = TRUE, exstr = "unMerged")
pairsDE1000vu <- getBestNFeatures(counts+0.5, clus.labels, "Pairs", 1000, isCount = TRUE, exstr = "unMerged")
pairsDEallvu <- getBestNFeatures(counts+0.5, clus.labels, "Pairs", Inf, isCount = TRUE, exstr = "unMerged")

##-- selecting unique DE genes
E3_unMerged_oneVall500V <- unique(oneVallDE500vu$Feature)
E3_unMerged_oneVallPairs500V <- unique(c(oneVallDE500vu$Feature,pairsDE500vu$Feature))


save(oneVallDEallvu, oneVallDE1000vu, oneVallDE500vu, file = file.path(DE_dir, paste0(esh, "_unMerged_oneVall_voomDE.Rda")))
save(oneVallDEallu, oneVallDE1000u, oneVallDE500u, file = file.path(DE_dir, paste0(esh, "_unMerged_oneVall_limmaDE.Rda")))
save(pairsDEallvu, pairsDE1000vu, pairsDE500vu, file = file.path(DE_dir, paste0(esh, "_unMerged_pairWise_voomDE.Rda")))
save(pairsDEallu, pairsDE1000u, pairsDE500u, file = file.path(DE_dir, paste0(esh, "_unMerged_pairWise_limmaDE.Rda")))
save(E3_unMerged_oneVall500, E3_unMerged_oneVallPairs500, file= file.path(DE_dir, paste0(esh, "_unMerged_limmaDEgeneLists.Rda")))
save(E3_unMerged_oneVall500V, E3_unMerged_oneVallPairs500V, file= file.path(DE_dir, paste0(esh, "_unMerged_voomDEgeneLists.Rda")))

```

### HBC3 DE
```{r HBC3_DE}
###---looking at merged cluster DE, cluster 11 = HBC3 (formerly iGBC)

cl11oneVallDE <- oneVallDEall[oneVallDEall$ContrastName == "11", ]
cl11Up <- cl11oneVallDE[cl11oneVallDE$logFC > 0, ]
cl11UpM1 <- cl11Up[cl11Up$logFC > 1,]
cl11UpM1 <- cl11UpM1[order(cl11UpM1$logFC, decreasing = TRUE), ]
cl11UpM2 <- cl11Up[cl11Up$logFC > 2,]
cl11UpM2 <- cl11UpM2[order(cl11UpM2$logFC, decreasing = TRUE), ]

save(cl11oneVallDE, cl11Up, cl11UpM1, cl11UpM2, file = file.path(DE_dir, paste0(esh, "_cl11_HBC3_DE.Rda")))

write.xlsx(cl11oneVallDE, file=file.path(DE_dir, paste0(esh, "cl11oneVallDE.xlsx")), sheetName="cl11oneVallDE")
write.xlsx(cl11Up, file=file.path(DE_dir, paste0(esh, "cl11oneVallDE.xlsx")), sheetName="cl11Up", append=TRUE)
write.xlsx(cl11UpM1, file=file.path(DE_dir, paste0(esh, "cl11oneVallDE.xlsx")), sheetName="cl11UpM1", append=TRUE)
write.xlsx(cl11UpM2, file=file.path(DE_dir, paste0(esh, "cl11oneVallDE.xlsx")), sheetName="cl11UpM2", append=TRUE)

```


```{r DEbyGenotype}
###---set-up
#####-----1. plan to subset the data to each cluster, then for each cluster, do DE between genotypes - may need to only do it on select clusters since some clusters have almost no Sox2cKO cells (neuronal lineage)
#####-----2. plan to do DE between genotype across all the data


deGeno <- function(counts, clusterVector, clusterID, isCount=FALSE, exstr=NULL){
  clMat <- counts[,clusterVector==clusterID]
  clgeno <- geno[colnames(clMat)]
  if(any(table(clgeno)==0)) return(NULL)
  adj <- ifelse(isCount, "voom", "limma")
  filename <- ifelse(is.null(exstr), paste0(esh, "_", adj, "_DE_expt.txt"), paste0(esh, "_", exstr, "_", adj, "_DE_expt.txt"))
  
  design <- model.matrix( ~ 0 + clgeno)
  if (isCount){
    dge <- DGEList(counts = clMat)
    v <- voom(dge, design)
    fit1 <<- lmFit(v, design)
  }else{
    fit1 <<- lmFit(clMat, design)  
  }
  
  con.matrix <- makeContrasts(clgenoK5ERRY - clgenoK5ERSOX2CKO, levels = design)
  fit2 <- contrasts.fit(fit1, con.matrix)
  fit2 <- eBayes(fit2)
  
  res <- topTable(fit2, coef = 1, adjust = "BH", number = Inf, p.value = 0.05)
  write.table(data.frame(cluster=rep(clusterID, nrow(res)),res), file = file.path(DE_dir, filename=filename), quote=FALSE, sep="\t", append=TRUE)
  de_genes <- rownames(res)
  return(de_genes)
}

###---make genotype vector
expt <- cmobjM@colData$expt
names(expt) <- colnames(cmobjM)
geno <- as.character(expt)
geno[grep("^K5ERSOX2CKO", geno)] <- "K5ERSOX2CKO"
geno[grep("^K5ERRY",geno)] <- "K5ERRY"
geno <- as.factor(geno)
names(geno) <- names(expt)

### remove files before running so we don't concatenate files endlessly
lapply(c("voom","limma"), function(adj){
file.remove(file.path(DE_dir, paste0(esh, "_", adj, "_DE_expt.txt")))
file.remove(file.path(DE_dir, paste0(esh, "_", "unMerged_", adj, "_DE_expt.txt")))
})

###---not using voom
de_limma <- lapply(setdiff(levels(clus.labelsM),-1), function(contrast) deGeno(counts_log, cluster=clus.labelsM, contrast, isCount=FALSE))
de_limmau <- lapply(setdiff(levels(clus.labels),-1), function(contrast) deGeno(counts_log, cluster=clus.labels, contrast, isCount=FALSE, exstr = "unMerged"))

###---using voom for DE
de_voom <- lapply(setdiff(levels(clus.labelsM),-1), function(contrast) deGeno(counts+1, cluster=clus.labelsM, contrast, isCount=TRUE))
de_voomu <- lapply(setdiff(levels(clus.labels),-1), function(contrast) deGeno(counts+1, cluster=clus.labels, contrast, isCount=TRUE))

cluster_ord <- c(12,5,11,3,6,9,15,14,4,1)

breakv <- c(min(counts_log), seq(0, quantile(counts_log[counts_log > 0], .99, na.rm = TRUE), length = 50), max(counts_log))

# lapply(list("de_limma", "de_voom"), function(de_name){
#   de <- unique(unlist(get(de_name)))
#   pdf(file = file.path(DE_dir, paste0(esh, '_hm_expt_',de_name,'.pdf')), width = 11, height = 8)
#   plotHeatmap(counts_log[de,], clusterSamples=FALSE, clusterFeatures=TRUE, breaks=breakv, sampleData=data.frame(clusters=clus.labelsM[order(match(clus.labelsM,cluster_ord))]), clusterLegend=list(clusters=col.pal))
#   dev.off()
# })
